File: frontend/src/types/product.ts
export interface Product {
    product_id: number;
    product_name: string;
    supplier_id: number | null;
    category_id: number | null;
    quantity_per_unit: string | null;
    unit_price: number | null;
    units_in_stock: number | null;
    units_on_order: number | null;
    reorder_level: number | null;
    discontinued: boolean;
    stock_status: 'in_stock' | 'low_stock' | 'out_of_stock' | 'discontinued';
    category: {
        category_id: number;
        category_name: string;
    } | null;
    supplier: {
        supplier_id: number;
        company_name: string;
        contact_name: string | null;
    } | null;
    created_at: string;
    updated_at: string;
}

export interface ProductListItem {
    product_id: number;
    product_name: string;
    category_name: string | null;
    supplier_name: string | null;
    unit_price: number | null;
    units_in_stock: number | null;
    stock_status: string;
    discontinued: boolean;
}

export interface ProductFilters {
    category_id?: number;
    supplier_id?: number;
    stock_status?: string;
    price_min?: number;
    price_max?: number;
    discontinued?: boolean;
}

export interface ProductListParams {
    page?: number;
    page_size?: number;
    search?: string;
    sort_by?: string;
    sort_order?: 'asc' | 'desc';
    category_id?: number;
    supplier_id?: number;
    stock_status?: string;
    price_min?: number;
    price_max?: number;
    discontinued?: boolean;
}

export interface ProductCreateInput {
    product_name: string;
    supplier_id?: number | null;
    category_id?: number | null;
    quantity_per_unit?: string | null;
    unit_price?: number | null;
    units_in_stock?: number | null;
    units_on_order?: number | null;
    reorder_level?: number | null;
    discontinued?: boolean;
}

export interface ProductUpdateInput {
    product_name?: string;
    supplier_id?: number | null;
    category_id?: number | null;
    quantity_per_unit?: string | null;
    unit_price?: number | null;
    units_in_stock?: number | null;
    units_on_order?: number | null;
    reorder_level?: number | null;
    discontinued?: boolean;
}

File: backend/app/schemas/product.py
from pydantic import BaseModel, Field, ConfigDict, field_validator
from typing import Optional, List
from datetime import datetime
from decimal import Decimal

class ProductBase(BaseModel):
    product_name: str = Field(..., min_length=1, max_length=100)
    supplier_id: Optional[int] = None
    category_id: Optional[int] = None
    quantity_per_unit: Optional[str] = Field(None, max_length=50)
    unit_price: Optional[Decimal] = Field(None, ge=0)
    units_in_stock: Optional[int] = Field(None, ge=0)
    units_on_order: Optional[int] = Field(None, ge=0)
    reorder_level: Optional[int] = Field(None, ge=0)
    discontinued: bool = False

class ProductCreate(ProductBase):
    pass

class ProductUpdate(BaseModel):
    product_name: Optional[str] = Field(None, min_length=1, max_length=100)
    supplier_id: Optional[int] = None
    category_id: Optional[int] = None
    quantity_per_unit: Optional[str] = None
    unit_price: Optional[Decimal] = Field(None, ge=0)
    units_in_stock: Optional[int] = Field(None, ge=0)
    units_on_order: Optional[int] = Field(None, ge=0)
    reorder_level: Optional[int] = Field(None, ge=0)
    discontinued: Optional[bool] = None

class CategoryInfo(BaseModel):
    category_id: int
    category_name: str
    model_config = ConfigDict(from_attributes=True)

class SupplierInfo(BaseModel):
    supplier_id: int
    company_name: str
    contact_name: Optional[str]
    model_config = ConfigDict(from_attributes=True)

class ProductResponse(ProductBase):
    product_id: int
    stock_status: str
    category: Optional[CategoryInfo] = None
    supplier: Optional[SupplierInfo] = None
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)

class ProductListResponse(BaseModel):
    product_id: int
    product_name: str
    category_name: Optional[str] = None
    supplier_name: Optional[str] = None
    unit_price: Optional[Decimal] = None
    units_in_stock: Optional[int] = None
    stock_status: str
    discontinued: bool
    
    model_config = ConfigDict(from_attributes=True)

    @field_validator('category_name', mode='before')
    @classmethod
    def get_category_name(cls, v, info):
        if hasattr(info.data.get('category'), 'category_name'):
            return info.data['category'].category_name
        return v

class ProductSummary(BaseModel):
    product_id: int
    product_name: str
    unit_price: Optional[Decimal]
    model_config = ConfigDict(from_attributes=True)
